cpp/src/arrow/dataset/file_base.cc

```diff
@@ -56,6 +57,32 @@ Result<std::shared_ptr<io::RandomAccessFile>> FileSource::Open() const {
   return custom_open_();
 }
 
+Result<std::shared_ptr<io::InputStream>> FileSource::OpenCompressed(
+    util::optional<Compression::type> compression) const {
+  ARROW_ASSIGN_OR_RAISE(auto file, Open());
+  auto actual_compression = Compression::type::UNCOMPRESSED;
+  if (!compression.has_value()) {
+    // Guess compression from file extension
```

Comment:
This can be made more compact with arrow::fs::internal::GetAbstractPathExtension and arrow::util::GetCompressionType